feature/phone number format e164

dan-netizen opened the pull request:
Add the option to format the phone number to E164 standard using the google libphonenumber library, which should be packaged as a dependency in the final jar for the authenticator to work.
Add the option to add the phone number to the 'mobile_number' user attribute on successful phone number validation action.
Based on authenticator configuration, if the phone does not pass the required formatting checks, an error is shown and the user is asked to re-enter a valid phone number.

svenseeberg commented:
> I could not find a way around making a jar with dependencies, which increases its size quite a bit.

The biggest issue here is that `io.quarkus` would be included, right? We only need this for development. If I understand https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#dependency-scope correctly, choosing the correct scope should mitigate the problem.

svenseeberg commented on sms-authenticator/pom.xml:
```diff
@@ -18,6 +18,11 @@
+		<dependency>
+			<groupId>com.googlecode.libphonenumber</groupId>
+			<artifactId>libphonenumber</artifactId>
+			<version>8.13.40</version>
```
We should probably activate Github Dependabot to automatically bump versions and receive security updates.

svenseeberg commented on sms-authenticator/src/main/resources/theme-resources/messages/messages_de.properties:
```diff
@@ -15,4 +15,8 @@ smsPhoneNumberInstructions=Bitte gib eine Telefonnummer inklusive L\u00E4ndervor
 smsPhoneUpdate=Empfange einen Einmal-Code per SMS auf dein Telefon.
 smsAuthenticator=SMS Authentifizierung
 
+numberFormatFailedToParse=Die Telefonnummer konnte nicht geparst werden. Bitte geben Sie es erneut ein.
```
```suggestion
numberFormatFailedToParse=Die Telefonnummer konnte nicht verarbeitet werden. Bitte versuchen Sie es erneut.
```

svenseeberg commented on sms-authenticator/src/main/resources/theme-resources/messages/messages_de.properties:
```diff
@@ -15,4 +15,8 @@ smsPhoneNumberInstructions=Bitte gib eine Telefonnummer inklusive L\u00E4ndervor
 smsPhoneUpdate=Empfange einen Einmal-Code per SMS auf dein Telefon.
 smsAuthenticator=SMS Authentifizierung
 
+numberFormatFailedToParse=Die Telefonnummer konnte nicht geparst werden. Bitte geben Sie es erneut ein.
+numberFormatNoMatchingFilters=Die Telefonnummer entspricht nicht den Filtern f�r den Telefonnummern Typ.
+numberFormatNumberInvalid=Ung�ltige Telefonnummer, bitte geben Sie sie erneut ein.
```
```suggestion
numberFormatFailedToParse=Die Telefonnummer konnte nicht verarbeitet werden. Bitte versuche es erneut.
numberFormatNoMatchingFilters=Diese Telefonnummer ist nicht gültig. Bitte nutze eine reguläre Mobilfunk- oder Festnetznummer.
numberFormatNumberInvalid=Ungültige Telefonnummer. Bitte versuche es erneut.
```

alexanderhofstaetter replied:
Better german translations are needed

melegiul replied:
Was addressed by fedcca05b59ba762cd5753df9aeb70f9d9efb188

svenseeberg commented on sms-authenticator/pom.xml:
```diff
@@ -18,6 +18,11 @@
+		<dependency>
+			<groupId>com.googlecode.libphonenumber</groupId>
+			<artifactId>libphonenumber</artifactId>
+			<version>8.13.40</version>
```
Just a general note: we should probably activate Github Dependabot to automatically bump versions and receive security updates.

melegiul replied:
Solved by 6922721b828f83d23627cbe16993c12a874b906f

dan-netizen commented:
I think I applied most of the suggestions, and replaced maven-assembly-plugin with maven-shade-plugin, that seemed to work better and the jar is now ~397KB down from ~13MB+.
Also added some other misc fixes and improvements.

alexanderhofstaetter commented:
There seems to be some problems regarding encoding on your side:

![image](https://github.com/netzbegruenung/keycloak-mfa-plugins/assets/6820505/52664eb8-383a-4e04-9c6a-4872371407bf)

melegiul commented on .github/dependabot.yml:
```diff
@@ -0,0 +1,15 @@
+      - dependency-name: org.apache.maven.plugins:maven-surefire-plugin
+      - dependency-name: org.apache.maven.plugins:maven-compiler-plugin
+      - dependency-name: io.quarkus:quarkus-maven-plugin
+      - dependency-name: org.apache.maven.plugins:maven-assembly-plugin
```
Maven assembly was replaced. I think in general it would make more sense to update the maven plugins and keycloak dependencies project wide. But that is out of scope for this PR.

So please just replace ignore by access `com.googlecode.libphonenumber:libphonenumber`.

melegiul reviewed (changes requested):
Thanks a lot for your contribution. Finally the long missing proper input validation. :heart_eyes: 

Just some minor issues, see below. Then this should be ready to merge.

alexanderhofstaetter commented on sms-authenticator/src/main/java/netzbegruenung/keycloak/authenticator/SmsAuthenticatorFactory.java:
```diff
@@ -86,7 +86,12 @@ public List<ProviderConfigProperty> getConfigProperties() {
+			new ProviderConfigProperty("mobileInputFieldPlaceholder", "Phone number input field placeholder", "The placeholder string user in the phone number input field", ProviderConfigProperty.STRING_TYPE, ""),
+			new ProviderConfigProperty("storeInAttribute", "Set phone number as attribute", "Sets the phone number as a user attribute.", ProviderConfigProperty.BOOLEAN_TYPE, false),
+			new ProviderConfigProperty("normalizePhoneNumber", "Format phone number", "Normalize the phone number using the E164 standard.", ProviderConfigProperty.BOOLEAN_TYPE, false),
+			new ProviderConfigProperty("numberTypeFilters", "Valid number type filters", "A comma separated list of valid number types.", ProviderConfigProperty.STRING_TYPE, ""),
```
Also we should put a link to all the possible values in the help icon text, so user known what to input here.

melegiul commented on sms-authenticator/pom.xml:
```diff
@@ -12,18 +12,25 @@
+		<dependency>
+			<groupId>com.googlecode.libphonenumber</groupId>
+			<artifactId>libphonenumber</artifactId>
+			<version>8.13.40</version>
 		</dependency>
 		<!-- Necessary for proper execution of IDELauncher -->
 		<!-- Can be removed as part of the https://github.com/keycloak/keycloak/issues/22455 enhancement -->
 		<dependency>
 			<groupId>io.quarkus</groupId>
 			<artifactId>quarkus-vertx-http-dev-ui-resources</artifactId>
-			<scope>provided</scope>
```
Please leave the scopes in sms-auth module.

In my original request I meant to remove scopes only from dependency management, which sits in the parent pom in the project root folder.

In summary:
- pom.xml -> no scopes
- sms-authenticator/pom.xml -> scopes

melegiul reviewed (changes requested):
Thank you very much for your changes. Just one thing left to fix. I guess just a misunderstanding. See below.

alexanderhofstaetter commented:
@melegiul can you provide a new release with all the changes?

melegiul commented:
Yes, sure. I will create a new release not later than tomorrow. I just want to add another feature, but unrelated to SMS.

melegiul commented:
And there it is :smiley: 
https://github.com/netzbegruenung/keycloak-mfa-plugins/releases/tag/v25.0.2
